src/VisualStudio/Next/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs

```diff
@@ -0,0 +1,178 @@
+                if (instance == null)
+                {
+                    // service is in shutdown mode or not enabled
+                    return null;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // make sure instance was started
+                return await instance.ConfigureAwait(false);
+            }
+
+            private async Task<RemoteHostClient> EnableAsync(CancellationToken cancellationToken)
+            {
+                await AddGlobalAssetsAsync(cancellationToken).ConfigureAwait(false);
+
+                return await StartInternalAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<RemoteHostClient> StartInternalAsync(CancellationToken cancellationToken)
+            {
+                // TODO: abstract this out so that we can have more host than service hub
+                var instance = await ServiceHubRemoteHostClient.CreateAsync(_workspace, cancellationToken).ConfigureAwait(false);
+                instance.ConnectionChanged += OnConnectionChanged;
+
+                return instance;
+            }
+
+            private async Task AddGlobalAssetsAsync(CancellationToken cancellationToken)
+            {
+                var snapshotService = _workspace.Services.GetService<ISolutionChecksumService>();
+                var assetBuilder = new AssetBuilder(_workspace.CurrentSolution);
+
+                foreach (var reference in _analyzerService.GetHostAnalyzerReferences())
+                {
+                    var asset = await assetBuilder.BuildAsync(reference, cancellationToken).ConfigureAwait(false);
+                    snapshotService.AddGlobalAsset(reference, asset, cancellationToken);
+                }
+            }
+
+            private void RemoveGlobalAssets()
+            {
+                var snapshotService = _workspace.Services.GetService<ISolutionChecksumService>();
+
+                foreach (var reference in _analyzerService.GetHostAnalyzerReferences())
+                {
+                    snapshotService.RemoveGlobalAsset(reference, CancellationToken.None);
+                }
+            }
+
+            // use local token and lock on instance
+            private void OnConnectionChanged(object sender, bool connection)
+            {
+                if (connection)
+                {
+                    return;
+                }
+
+                // if remote host gets disconnected. tell users that remote host is gone and whether they want to recover remote host.
+                lock (_gate)
+                {
+                    if (_shutdown.IsCancellationRequested)
+                    {
+                        // we are shutting down.
+                        return;
+                    }
+
+                    _instance = null;
+                }
+
+                _workspace.Services.GetService<IErrorReportingService>().ShowErrorInfo(
```

Comment:
are  you talking about IErrorReportingService? it is free thread service, it will take care of calling VS API from ui thread.